### PR TITLE
[LoRA] support kohya and xlabs loras for flux.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Fixes https://github.com/huggingface/diffusers/issues/9291. 
Fixes https://github.com/huggingface/diffusers/issues/9292 (potentially). 

@apolinario does this work for you? 

I decided to reuse https://github.com/kohya-ss/sd-scripts/blob/sd3/networks/convert_flux_lora.py (with appropriate credits to @kohya-ss) because the conversion utility won't have been too different if we were to write it anyway. 

@kohya-ss could I get your commit email address so that we could add you as a co-author of this PR? @Leommm-byte same for you as well. 

About tests, I think it might make sense to add a bit later to give the community a chance to play with it.

### CODE REVIEW

**Comment by apolinario** [comment]
The Kohya LoRA seem to be working! 
```py
import torch
from diffusers import DiffusionPipeline

device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = DiffusionPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16)
pipe = pipe.to(device)

pipe.load_lora_weights("Norod78/brain-slug-flux")
generator = torch.Generator(device="cpu").manual_seed(0)
prompt = "The girl with a brain slug earring"
image = pipe(prompt, generator=generator).images[0]
image
```
download copy

**Comment by sayakpaul** [comment]
Sure, anyway, I have made a comment about the code I reused from `sd-scripts`. Thank you!

**Comment by apolinario** [comment]
XLabs LoRAs seem to be working, however they seem to need a higher scale to really shine/showcase their effect (also, using the guidance_scale of `3.5` instead of the default `7.0`, as the super high guidance seems to outshine the effect of the LoRA) 

<table>
  <tr>
    <td colspan="3" align="center">A blue jay standing on a large basket of rainbow macarons, disney style (<a href="https://huggingface.co/XLabs-AI/flux-lora-collection/blob/main/disney_lora.safetensors">disney_lora.safetensors</a>)</td>
  </tr>
  <tr>
    <th>No LoRA</th>
    <th>disney_lora.safetensors (scale 1.0)</th>
    <th>disney_lora.safetensors (scale 1.75)</th>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

<table>
  <tr>
    <td colspan="3" align="center">The portrait of a brazilian person (<a href="https://huggingface.co/XLabs-AI/flux-RealismLora">XLabs-AI/flux-RealismLora</a>)</td>
  </tr>
  <tr>
    <th>No LoRA</th>
    <th>flux-RealismLora (scale 1.0)</th>
    <th>flux-RealismLora (scale 1.75)</th>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>


```py
import torch
from diffusers import DiffusionPipeline

device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = DiffusionPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16)
pipe = pipe.to(device)

prompt = "A blue jay standing on a large basket of rainbow macarons, disney style"
#prompt = "The portrait of a brazilian person"

generator = torch.Generator(device="cpu").manual_seed(42)
image = pipe(prompt, generator=generator, guidance_scale=3.5).images[0]
image.save("no_lora.png")

pipe.load_lora_weights("XLabs-AI/flux-lora-collection", weight_name="disney_lora.safetensors")
#pipe.load_lora_weights("XLabs-AI/flux-RealismLora")

generator = torch.Generator(device="cpu").manual_seed(42)
image = pipe(prompt, generator=generator, joint_attention_kwargs={"scale": 1}, guidance_scale=3.5).images[0]
image.save("lora_1_0.png")

generator = torch.Generator(device="cpu").manual_seed(42)
image = pipe(prompt, generator=generator, joint_attention_kwargs={"scale": 1.75}, guidance_scale=3.5).images[0]
image.save("lora_1_75.png")
```

**Comment by sayakpaul** [comment]
> One other thing, some LoRAs in this format have a diffusion_model prepend before the double_blocks, as in

@apolinario thanks! Does d3c60e537 work for you?

File: src/diffusers/loaders/lora_conversion_utils.py [hunk 1]
```diff
@@ -326,3 +328,294 @@ def _get_alpha_name(lora_name_alpha, diffusers_name, alpha):
...
+
+        # down_weight is copied to each split
+        ait_sd.update({k: down_weight for k in ait_down_keys})
+
+        # up_weight is split to each split
+        ait_sd.update({k: v for k, v in zip(ait_up_keys, torch.split(up_weight, dims, dim=0))})  # noqa: C416
+
+    for old_key in orig_keys:
+        # Handle double_blocks
+        if old_key.startswith(("diffusion_model", "double_blocks")):
```
Comment (apolinario) [inline src/diffusers/loaders/lora_conversion_utils.py:557]: Not sure if all LoRAs that start with `diffusion_model` will have the `.double_blocks` suffix. Starting with `diffusion_model` seem to be a COmfyUI compatibility thing. Maybe this? 

```suggestion
        if old_key.startswith(("diffusion_model.double_blocks", "double_blocks")):
```

File: src/diffusers/loaders/lora_conversion_utils.py [hunk 2]
```diff
@@ -326,3 +328,294 @@ def _get_alpha_name(lora_name_alpha, diffusers_name, alpha):
...
+                )
+                # continue
+
+            if "down" in old_key:
+                new_key += ".lora_A.weight"
+            elif "up" in old_key:
+                new_key += ".lora_B.weight"
+
+        # Handle single_blocks
+        elif old_key.startswith("diffusion_model", "single_blocks"):
```
Comment (apolinario) [inline src/diffusers/loaders/lora_conversion_utils.py:596]:
```suggestion
        elif old_key.startswith("diffusion_model.single_blocks", "single_blocks"):
```
Same suggestion

File: src/diffusers/loaders/lora_pipeline.py [hunk 3]
```diff
@@ -1583,6 +1588,20 @@ def lora_state_dict(
...
 
+        # TODO (sayakpaul): to a follow-up to clean and try to unify the conditions.
+
+        is_kohya = any(".lora_down.weight" in k for k in state_dict)
+        if is_kohya:
+            state_dict = _convert_kohya_flux_lora_to_diffusers(state_dict)
+            # Kohya already takes care of scaling the LoRA parameters with alpha.
+            return (state_dict, None) if return_alphas else state_dict
+
+        is_xlabs = any("processor" in k for k in state_dict)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/loaders/lora_pipeline.py:1599]: I think not. 

https://huggingface.co/Norod78/brain-slug-flux/tree/main?show_file_info=Brain_Slug_FLUX.safetensors also `double_blocks`. I'd assume all the loRA trainers basing their implementations on the original BFL codebase would have that. So, we need to be a bit more careful.

[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/loaders/lora_pipeline.py:1599]: Well the key doesn't start with `double_blocks` per se, rather it starts with `lora_unet_double_blocks`, no? But yes, I understand your point. Maybe we can keep `processor`, if LoRAs in the future with `processor` and other format emerge we can act

[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/loaders/lora_pipeline.py:1599]: I meant substring. Not `startswith` particularly. Modified my comment.

**Review by apolinario — APPROVED** [review]
Let's go!

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by duonglegiang** [comment]
Do you have plans to support ControlNet and IP Adapter from XLabs-AI in future updates?
ControlNet: https://huggingface.co/XLabs-AI/flux-controlnet-canny
IP-Adapter: https://huggingface.co/XLabs-AI/flux-ip-adapter

**Comment by sayakpaul** [comment]
For ControlNet, https://github.com/huggingface/diffusers/issues/9301. 

For IP-Adapters, cc: @fabiorigano  @asomoza @yiyixuxu.

**Comment by Oguzhanercan** [comment]
High, thanks for your work. I wonder that, using trained loras-controlnets with 4-bit quantized DiTs(either nf4, raw int4 or SVD4Bit) is okey? Or should we quantize them also? As I see, flux-labs scripts are using float32 dtype. Loading and fusing them to 4bit dit is not throws an error. But I am not sure that loras trained with these scripts works with diffusers beacuse the output does not change at all.


@sayakpaul

**Comment by sayakpaul** [comment]
You need to keep some things in mind when fusing LoRAs into a quantized base model:
https://github.com/huggingface/diffusers/tree/main/examples/research_projects/flux_lora_quantization#inference

**Comment by Oguzhanercan** [comment]
> You need to keep some things in mind when fusing LoRAs into a quantized base model:
> 
> https://github.com/huggingface/diffusers/tree/main/examples/research_projects/flux_lora_quantization#inference

Thank you

**Comment by dancasas** [comment]
I am using the provided code example to load a LoRA I trained with X-Labs, but there seems to be some new keys that the current implementation of [lora_conversion_utils.py](https://github.com/huggingface/diffusers/pull/9295/files/d3c60e5370ceb035746d336bc48e249af02200c5#diff-8dfd6649c3ba96b8610f767df56209ba78d6ea3a4a1db6f510b919a9dd6dd710) does not parse. I get this error after `pipe.load_lora_weights("/tmp/my_lora.safetensors")`

> ValueError: `old_state_dict` should be at this point but has: ['single_blocks.1.processor.qkv_lora.down.weight', 'single_blocks.1.processor.qkv_lora.up.weight', 'single_blocks.2.processor.qkv_lora.down.weight', 'single_blocks.2.processor.qkv_lora.up.weight', 'single_blocks.3.processor.qkv_lora.down.weight', 'single_blocks.3.processor.qkv_lora.up.weight', 'single_blocks.4.processor.qkv_lora.down.weight', 'single_blocks.4.processor.qkv_lora.up.weight'].

Does anyone know how to fix it? Thanks!

**Comment by sayakpaul** [comment]
Refer to https://github.com/huggingface/diffusers/pull/9915.

**Comment by dancasas** [comment]
Thanks! I can confirm that this PR https://github.com/huggingface/diffusers/pull/9915 worked for me to :)

**Pull request merged**